tools/data/build_audio_features.py

```diff
@@ -0,0 +1,265 @@
+        spectro_mag = np.expand_dims(spectro_mag, axis=0)
+        if with_phase:
+            spectro_phase = np.expand_dims(np.angle(spectro_phase), axis=0)
+            return spectro_mag, spectro_phase
+        else:
+            return spectro_mag
+
+    def save_wav(self, wav, path):
+        """Save the wav to disk."""
+        wav *= 32767 / max(0.01, np.max(np.abs(wav)))
+        wavfile.write(path, self.sample_rate, wav.astype(np.int16))
+
+    def trim(self, quantized):
+        """Trim the audio wavfile."""
+        start, end = self.start_and_end_indices(quantized,
+                                                self.silence_threshold)
+        return quantized[start:end]
+
+    def adjust_time_resolution(self, quantized, mel):
+        """Adjust time resolution by repeating features.
+
+        Args:
+            quantized (ndarray): (T,)
+            mel (ndarray): (N, D)
+
+        Returns:
+            tuple: Tuple of (T,) and (T, D)
+        """
+        assert len(quantized.shape) == 1
+        assert len(mel.shape) == 2
+
+        upsample_factor = quantized.size // mel.shape[0]
+        mel = np.repeat(mel, upsample_factor, axis=0)
+        n_pad = quantized.size - mel.shape[0]
+        if n_pad != 0:
+            assert n_pad > 0
+            mel = np.pad(
+                mel, [(0, n_pad), (0, 0)], mode='constant', constant_values=0)
+
+        # trim
+        start, end = self.start_and_end_indices(quantized,
+                                                self.silence_threshold)
+
+        return quantized[start:end], mel[start:end, :]
+
+    def start_and_end_indices(self, quantized, silence_threshold=2):
+        """Trim the audio file when reaches the silence threshold."""
+        for start in range(quantized.size):
+            if abs(quantized[start] - 127) > silence_threshold:
+                break
+        for end in range(quantized.size - 1, 1, -1):
+            if abs(quantized[end] - 127) > silence_threshold:
+                break
+
+        assert abs(quantized[start] - 127) > silence_threshold
+        assert abs(quantized[end] - 127) > silence_threshold
+
+        return start, end
+
+    def melspectrogram(self, y):
+        """Generate the melspectrogram."""
+        D = self._lws_processor().stft(y).T
+        S = self._amp_to_db(self._linear_to_mel(np.abs(D))) - self.ref_level_db
+        if not self.allow_clipping_in_normalization:
+            assert S.max() <= 0 and S.min() - self.min_level_db >= 0
+        return self._normalize(S)
+
+    def get_hop_size(self):
+        """Calculate the hop size."""
+        hop_size = self.hop_size
+        if hop_size is None:
+            assert self.frame_shift_ms is not None
+            hop_size = int(self.frame_shift_ms / 1000 * self.sample_rate)
+        return hop_size
+
+    def _lws_processor(self):
+        return lws.lws(self.fft_size, self.get_hop_size(), mode='speech')
+
+    def lws_num_frames(self, length, fsize, fshift):
```

Comment:
Detailed args description are needed. (`fsize` and `fshift` are confused)

Reply:
There are many signal related terms that I am also not familiar with, I just follow the official instruction https://pypi.org/project/lws/1.2.6/. and deepvoice3 https://github.com/r9y9/deepvoice3_pytorch. Maybe add a reference link to them?

Reply:
Yeah, it is good to add a reference link